### PR TITLE
Recovery certificate repository (EXPOSUERAPP-7617)

### PR DESCRIPTION
**Pull request by chris-cwa**
This PR covers storage of recovery certificates. On review please check wether the correct identifiers are being compared.

- [ ] unit testing (in follow-up PR)
- [x] use container id from https://github.com/corona-warn-app/cwa-app-android/pull/3504

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    @BaseGson val baseGson: Gson,
+) {
+
+    private val prefs by lazy {
+        context.getSharedPreferences("recovery_localdata", Context.MODE_PRIVATE)
+    }
+
+    private val gson by lazy {
+        baseGson.newBuilder().create()
+    }
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt:25]: why do we need this? can't we just use `baseGson` directly ?

**Review by mtwalli — COMMENTED** [review]
Lgtm

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            prefs.edit {
+                prefs.all.keys.filter { it.startsWith(PKEY_RECOVERY_CERT_PREFIX) }.forEach {
+                    Timber.tag(TAG).v("Removing data for %s", it)
+                    remove(it)
+                }
+                value.forEach {
+                    val raw = gson.toJson(it)
+                    val identifier = it.identifier
+                    Timber.tag(TAG).v("Storing recovery certificate %s -> %s", identifier, raw)
+                    putString("$PKEY_RECOVERY_CERT_PREFIX$identifier", raw)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt:54]: What will `identifier` be? A UUID?
Alternative would be store them not individually but as a `Set<>`, I'm not quite sure what would be better in this case :thinking:

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt [hunk 3]
```diff
@@ -20,23 +29,64 @@ class RecoveryCertificateRepository @Inject constructor(
...
+                    qrCodeExtractor = qrCodeExtractor
+                )
+            }
+            .toSet()
+            .also { Timber.tag(TAG).v("Restored recovery certificate data: %s", it) }
+    }
+
+    init {
+        internalData.data
+            .onStart { Timber.tag(TAG).d("Observing test data.") }
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:53]: `test` typo?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt [hunk 4]
```diff
@@ -20,23 +29,64 @@ class RecoveryCertificateRepository @Inject constructor(
...
+            .catch {
+                it.reportProblem(TAG, "Failed to snapshot recovery certificate data to storage.")
+                throw it
+            }
+            .launchIn(appScope + dispatcherProvider.IO)
+    }
+
+    val certificates: Flow<Set<RecoveryCertificateWrapper>> = combine(
+        internalData.data,
+        valueSetsRepository.latestVaccinationValueSets // TODO use corrects sets
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:67]: According to @BMItter there will be no valuesets for recovery certificates, at the moment.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt [hunk 5]
```diff
@@ -20,23 +29,64 @@ class RecoveryCertificateRepository @Inject constructor(
...
         Timber.tag(TAG).d("requestCertificate(qrCode=%s)", qrCode)
         throw NotImplementedError()
     }
 
     suspend fun deleteCertificate(identifier: RecoveryCertificateIdentifier) {
-        Timber.tag(TAG).d("deleteTestCertificate(identifier=%s)", identifier)
-        throw NotImplementedError()
+        Timber.tag(TAG).d("deleteCertificate(identifier=%s)", identifier)
+        internalData.updateBlocking {
+            mapNotNull { if (it.certificateId == identifier) null else it }.toSet()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:80]: `certificateId` is different from `identifier`

* `certificateId` is the ID field in the scanned certificate JSON
* `identifier` is a unique ID we come up with

We could make them the same, but we didn't do this for the other certificates as we would have to rely on them being unique (which I don't think we should do, as we have seen, countries do all sorts of different things that are not according to schema specs...).

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt (hunk 4 above)
Comment (vaubaehn) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:67]: They just don't care about any vaccine or test detail ... currently. Just stating who issued where, and how long valid:
https://github.com/ehn-dcc-development/ehn-dcc-schema/blob/a603410d760fefc9073931c8c807759d9714c136/DCC.combined-schema.json#L267

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt (hunk 1 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt:24]: thank you!

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt (hunk 4 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:67]: removed it

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt (hunk 5 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:76]: introduced uuid for client side identification

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt [hunk 6]
```diff
@@ -1,42 +1,88 @@
...
+    }
+
+    val certificates: Flow<Set<RecoveryCertificateWrapper>> =
+        internalData.data.transform { set ->
+            set.map { RecoveryCertificateWrapper(null, it) }.toSet()
+        }
 
     suspend fun requestCertificate(qrCode: RecoveryCertificateQRCode): RecoveryCertificateContainer {
         Timber.tag(TAG).d("requestCertificate(qrCode=%s)", qrCode)
         throw NotImplementedError()
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:70]: is that part of this PR ? or another one ?

Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt]: pr was draft only

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/StoredRecoveryCertificateData.kt [hunk 7]
```diff
@@ -7,7 +7,18 @@ data class StoredRecoveryCertificateData(
...
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as StoredRecoveryCertificateData
+
+        if (identifier != other.identifier) return false
+
+        return true
+    }
+}
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/StoredRecoveryCertificateData.kt:21]: Why the custom `equals` here?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        set(value) {
+            Timber.tag(TAG).d("recoveryCertificates - save(%s)", value)
+            prefs.edit()
+                .putStringSet(
+                    PKEY_RECOVERY_CERT,
+                    value.map { data ->
+                        gson.toJson(data).also {
+                            Timber.tag(TAG).v("Storing recovery certificate %s -> %s", data.identifier, it)
+                        }
+                    }.toSet()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt:42]: What do you think about storing the whole set as json vs using `putStringSet`? `putStringSet` has a spotty history, with a few footguns because it doesn't return a copy of the internal set when using `get`.

e.g.
```kotlin
      val type =  object : TypeToken<Set<StoredRecoveryCertificateData>>() {}.type
      gson.toJson<Set<StoredRecoveryCertificateData>>(data, type)
```

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                )
+                .apply()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt:44]: Could use `edit { }` instead of `edit().....apply()`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+                gson.fromJson<StoredRecoveryCertificateData>(it).also { data ->
+                    Timber.tag(TAG).v("recovery certificate loaded: %s", data)
+                    requireNotNull(data.identifier)
+                    requireNotNull(data.registeredAt)
+                }
+            }?.toSet() ?: emptySet()
+        }
+        set(value) {
+            Timber.tag(TAG).d("recoveryCertificates - save(%s)", value)
+            prefs.edit()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt:35]: Other storages we have remove the key when the set is empty, here we write an empty set, should we align?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/DuplicateRecoveryCertificateException.kt [hunk 11]
```diff
@@ -0,0 +1,5 @@
+package de.rki.coronawarnapp.covidcertificate.recovery.core
+
+class DuplicateRecoveryCertificateException(
+    message: String
+) : IllegalArgumentException(message)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/DuplicateRecoveryCertificateException.kt:5]: Please use the existing:
```kotlin
InvalidRecoveryCertificateException(
    InvalidHealthCertificateException.ErrorCode.ALREADY_REGISTERED
)
```

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt [hunk 12]
```diff
@@ -1,42 +1,108 @@
...
+        internalData.data.transform { set ->
+            set.map { RecoveryCertificateWrapper(null, it) }.toSet()
+        }
+
+    @Throws(DuplicateRecoveryCertificateException::class)
     suspend fun requestCertificate(qrCode: RecoveryCertificateQRCode): RecoveryCertificateContainer {
         Timber.tag(TAG).d("requestCertificate(qrCode=%s)", qrCode)
-        throw NotImplementedError()
+        qrCodeExtractor.extract(qrCode.qrCode).toContainer().apply {
+            if (internalData.data.last().contains(this)) {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:75]: This hangs forever. `last()` waits for the terminal operator of a flow, but `internalData.data` never completes in normal operation. You likely want `first()` here to get the newest value.

Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:75]: But then we are left with a racecondition. Between performing the check here, and updating the internal data with `updateBlocking`, the data may have changed and we could be adding a duplicate certificate. Unlikely, but still...

Place the check inside `updateBlocking { currentData -> .... }` and perform it against `currentData` to prevent the racecondition.

**Review by d4rken — DISMISSED** [review]
Right approach, few bugs.

Please add some tests for the duplicate certificate exception and the json storage format.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/StoredRecoveryCertificateData.kt (hunk 7 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/StoredRecoveryCertificateData.kt:21]: because there was no container id yet :)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/DuplicateRecoveryCertificateException.kt (hunk 11 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/DuplicateRecoveryCertificateException.kt:5]: using it now

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt (hunk 8 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/storage/RecoveryCertificateStorage.kt]: created a dto for json

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt [hunk 13]
```diff
@@ -1,42 +1,110 @@
...
+    val certificates: Flow<Set<RecoveryCertificateWrapper>> =
+        internalData.data.transform { set ->
+            set.map { RecoveryCertificateWrapper(null, it) }.toSet()
+        }
+
+    @Throws(InvalidRecoveryCertificateException::class)
     suspend fun registerCertificate(qrCode: RecoveryCertificateQRCode): RecoveryCertificateContainer {
         Timber.tag(TAG).d("registerCertificate(qrCode=%s)", qrCode)
-        throw NotImplementedError()
+        return qrCodeExtractor.extract(qrCode.qrCode).toContainer().apply {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:75]: I don't think we need to extract the qrCode again here, `qrCodeExtractor.extract(qrCode.qrCode)` basically returns `qrCode: RecoveryCertificateQRCode` which we already have as function parameter.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt [hunk 14]
```diff
@@ -1,42 +1,110 @@
...
+            internalData.updateBlocking {
+                if (any { it.containerId == containerId }) {
+                    throw InvalidRecoveryCertificateException(InvalidHealthCertificateException.ErrorCode.ALREADY_REGISTERED)
+                }
+                plus(this)
+            }
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:81]: I think it's the wrong `this`, `this` in that scope might be the current data snapshot from `updateBlocking`?

```kotlin
fun updateBlocking(
    action: suspend Set<RecoveryCertificateContainer>.() -> Set<RecoveryCertificateContainer>
): Set<RecoveryCertificateContainer>
```

`plus(this)` would be set + set, duplicating the original data and ignoring the new value.
So no compiler error because `SET.plus(SET)` is also valid.

You likely meant `this@apply`, but maybe we can write this a little bit different so future devs don't accidentally break this without noticing.

Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt]: put the new container into new var

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt (hunk 13 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt]: thank you

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt [hunk 15]
```diff
@@ -1,42 +1,111 @@
...
+            plus(newContainer)
+        }
+        return newContainer
     }
 
-    suspend fun deleteCertificate(containerId: RecoveryCertificateContainerId): RecoveryCertificateContainer? {
+    private fun RecoveryCertificateQRCode.toContainer() = RecoveryCertificateContainer(
+        data = StoredRecoveryCertificateData(
+            identifier = uniqueCertificateIdentifier,
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:88]: If the identifier is the `uniqueCertificateIdentifier` from the certificate, then we don't need to store it extra.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt [hunk 16]
```diff
@@ -1,42 +1,111 @@
...
+            set.map { RecoveryCertificateWrapper(null, it) }.toSet()
+        }
+
+    @Throws(InvalidRecoveryCertificateException::class)
     suspend fun registerCertificate(qrCode: RecoveryCertificateQRCode): RecoveryCertificateContainer {
         Timber.tag(TAG).d("registerCertificate(qrCode=%s)", qrCode)
-        throw NotImplementedError()
+        val newContainer = qrCode.toContainer()
+        internalData.updateBlocking {
+            if (any { it.containerId == newContainer.containerId }) {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:76]: Prefer checking here for `uniqueCertificateIdentifier` explicitly, as what `containerId` is backed by may change over the time.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt [hunk 17]
```diff
@@ -1,42 +1,111 @@
...
+            plus(newContainer)
+        }
+        return newContainer
     }
 
-    suspend fun deleteCertificate(containerId: RecoveryCertificateContainerId): RecoveryCertificateContainer? {
+    private fun RecoveryCertificateQRCode.toContainer() = RecoveryCertificateContainer(
+        data = StoredRecoveryCertificateData(
+            identifier = uniqueCertificateIdentifier,
+            registeredAt = data.header.issuedAt,
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:89]: `registeredAt` (in other places in the app) refers to the time in the app when the code was scanned. We should use a similar timestamp or rename the property (but if we rename the property, we don't need to store the headerfield twice, as we store the whole qrcode including header).

**Comment by harambasicluka** [comment]
As @d4rken is on vacation someone else has to take over the review.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt (hunk 17 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:89]: What is the field registeredAt needed for? I cannot find any usage

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/DuplicateRecoveryCertificateException.kt (hunk 11 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/DuplicateRecoveryCertificateException.kt:5]: Please delete the exception

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt [hunk 18]
```diff
@@ -1,42 +1,109 @@
...
+import kotlinx.coroutines.flow.transform
+import kotlinx.coroutines.plus
 import timber.log.Timber
 import javax.inject.Inject
 import javax.inject.Singleton
 
 @Singleton
 class RecoveryCertificateRepository @Inject constructor(
     @AppScope private val appScope: CoroutineScope,
-    private val dispatcherProvider: DispatcherProvider,
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:20]: Why is it not private any more?

Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt:20]: i removed `val` also - so it's not a class member anymore - it is only used in `init`

**Review by chiljamgossow — DISMISSED** [review]
Let's merge and test the whole flow

**Comment by mtwalli** [comment]
@chris-cwa  I don't see any certificates in your branch.  Comparing to `2.5` where Test and vaccination certificates are showing up after scanning. Here I can not see them anymore , Could you please have a look ?

**Review by mtwalli — APPROVED** [review]
We can merge , I just pushed a fix for flow emission

**Review by chiljamgossow — APPROVED** [review]
